File: java/src/main/java/com/haibo/yan/algorithm/dp/EditDistance.java
/*
 * Copyright (c) [2017] [Haibo(Tristan) Yan]
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.haibo.yan.algorithm.dp;

public class EditDistance {
    /**
     * see https://leetcode.com/problems/edit-distance
     * Given two words word1 and word2, find the minimum number of steps required to convert word1 to word2.
     * (each operation is counted as 1 step.)
     * You have the following 3 operations permitted on a word:
     * a) Insert a character
     * b) Delete a character
     * c) Replace a character
     * @param word1
     * @param word2
     * @return
     */
    public int minDistance(String word1, String word2) {
        char[] wc1 = word1.toCharArray(), wc2 = word2.toCharArray();

        int[][] distances = new int[wc1.length + 1][wc2.length + 1];

        for (int i = 0; i <= wc1.length; i++) {
            distances[i][0] = i;
        }

        for (int j = 0; j <= wc2.length; j++) {
            distances[0][j] = j;
        }

        for (int i = 1; i <= wc1.length; i++) {
            for (int j = 1; j <= wc2.length; j++) {
                char x = wc1[i - 1], y = wc2[j - 1];
                if (x == y) {
                    distances[i][j] = distances[i - 1][j - 1];
                } else {
                    distances[i][j] = Math.min(distances[i - 1][j], distances[i][j - 1]) + 1;
                    distances[i][j] = Math.min(distances[i - 1][j - 1] + 1, distances[i][j]);
                }
            }
        }

        return distances[wc1.length][wc2.length];
    }

    /**
     * Given two strings S and T, determine if they are both one edit distance apart.
     */
    public boolean isOneEditDistance(String s, String t) {
        char[] sc = s.toCharArray(), tc = t.toCharArray();

        int i = 0,  len = Math.min(sc.length, tc.length);

        int distance = tc.length - len + sc.length - len;

        while (i < len) {
            if (sc[i] != tc[i]) {
                distance++;
            }
        }

        return distance == 1;
    }
}

File: c++/algorithm/dp/EditDistance.cpp
//Copyright [2018] [Haibo Yan]
//
//Licensed under the Apache License, Version 2.0 (the "License");
//you may not use this file except in compliance with the License.
//You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
//Unless required by applicable law or agreed to in writing, software
//distributed under the License is distributed on an "AS IS" BASIS,
//WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//See the License for the specific language governing permissions and
//limitations under the License.

//
// Created by Haibo Yan on 4/5/18.
//

#include "EditDistance.h"

int EditDistance::minD(string s1, string s2) {
    int d[s1.size()][s2.size()];

    for (int i = 0; i < s1.size(); i++) {
        d[i][0] = s1[i] == s2[0] ? 0: 1;
    }
    for (int j = 0; j < s2.size(); j++) {
        d[0][j] = s1[0] == s2[j] ? 0: 1;
    }

    for (int i = 1; i < s1.size(); i++) {
        for (int j = 0; j < s2.size(); j++) {
            if (s1[i] == s2[j]) {
                d[i][j] = d[i - 1][j - 1];
            } else {
                d[i][j] = min(min(d[i - 1][j], d[i - 1][j - 1]), d[i][j - 1]) + 1;
            }
        }
    }

    return d[s1.size() - 1][s2.size() - 1];
}

bool EditDistance::oneEditDistance(string s1, string s2) {
    int d = s1.size() - s2.size();

    if (d < -1 || d > 1) {
        return false;
    }

    for (int i = 0; i < min(s1.size(), s2.size()); i++) {
        if (s1[i] != s2[i]) {
            if (d == 0) {
                return equals(s1, s2, i + 1, i + 1, s1.size() - i - 1);
            } else if (d == -1) {
                return equals(s1, s2, i, i + 1, s1.size() - i);
            } else {
                return equals(s1, s2, i + 1, i, s2.size() - i);
            }
        }
    }

    return true;
}

bool EditDistance::equals(string s1, string s2, int b1, int b2, int l) {
    int e1 = b1 + l, e2 = b2 + l;
    if (e1 > s1.size() || e2 > s2.size()) {
        return false;
    }

    for (int i = 0; i < l; i++) {
        if (s1[i + b1] != s2[i + b2]) {
            return false;
        }
    }

    return true;
}